#166272337 Forgot Password

EMacco opened the pull request:
#### What does this PR do?
Allows user to be able to reset password if it was forgotten

#### Description of Task to be completed?
- Write unit tests
- Document feature using swagger
- Create route `POST auth/reset_password?token=&email=`
- Create route `POST auth/forgot_password`
- Generate and hash reset token
- Setup email template
- Validate new Password
- Send reset and success email

#### How should this be manually tested?
- pull this branch
- run "npm install"
- run "npm test"

#### Any background context you want to provide?
N/A

#### What are the relevant pivotal tracker stories?
[#166272337](https://www.pivotaltracker.com/story/show/166272337)

#### Screenshots (if appropriate)
![IMG_9821](https://user-images.githubusercontent.com/20377428/59839794-1e320100-9349-11e9-873a-3e2fc0c794f4.PNG)
![IMG_9822](https://user-images.githubusercontent.com/20377428/59839800-20945b00-9349-11e9-8172-40a1b9508b04.PNG)



#### Questions:
N/A

EMacco committed:
feat-[166272337]: Forgot Password  …
- Write unit tests
- Document feature using swagger
- Create route auth/reset_password?token=&email=
- Create route auth/forgot_password
- Generate and hash reset token
- Setup email template
- Validate new Password
- Send reset and success email

[Finishes #166272337]

EMacco committed:
Merge branch 'develop' into 166272337-ft-reset-password

EMacco committed:
Merge branch 'develop' into 166272337-ft-reset-password

EMacco committed:
Merge branch 'develop' into 166272337-ft-reset-password

fidelisojeah commented on server/controllers/users.js:
```diff
@@ -138,20 +162,124 @@ class UserController {
+      const token = randomString({ length: 40 });
+      const resetDetails = {
+        resetPasswordToken: token,
+        resetPasswordExpiry: Date.now() + 3600000 // 1 hour from now
```
too short (1 day is good), however, it's better this value is not hardcode but gotten from an environment variable (with a default value supplied so things don't break)

fidelisojeah commented on server/controllers/users.js:
```diff
@@ -138,20 +162,124 @@ class UserController {
+      if (!user) return Response.error(res, 404, 'User does not exist');
+
+      // Check if the account is active
+      if (!user.active) return Response.error(res, 401, 'Your account is not verified');
```
403 for this though

fidelisojeah reviewed (changes requested):
Nice job with the email design.

Consider making it some sort of template so others could use parts of it that are similar to what they need.

EMacco committed:
Merge branch '166272337-ft-reset-password' of https://github.com/andela/kingsmen-ah-backend into 166272337-ft-reset-password

EMacco committed:
Merge branch '166272337-ft-reset-password' of https://github.com/andela/kingsmen-ah-backend into 166272337-ft-reset-password

fidelisojeah commented on server/tests/factory/user-factory.js:
```diff
@@ -11,14 +11,14 @@ const generateToken = async (userDetails) => {
     email: email || faker.internet.email(),
     password: password || faker.internet.password(),
-    active: true
+    active: active || false
```
No please,

create an extra factory function for non active users and use that for your tests.

EMacco committed:
feat-[166272337]: Forgot Password

- Write unit tests
- Document feature using swagger
- Create route auth/reset_password
- Create route auth/forgot_password
- Generate and hash reset token
- Setup email template using mailgen
- Validate new Password
- Send reset and success emails

[Finishes #166272337]

EMacco committed:
feat-[166272337]: Forgot Password

- Write unit tests
- Document feature using swagger
- Create route auth/reset_password
- Create route auth/forgot_password
- Generate and hash reset token
- Setup email template using mailgen
- Validate new Password
- Send reset and success emails

[Finishes #166272337]

EMacco committed:
feat-[166272337]: Forgot Password

- Write unit tests
- Document feature using swagger
- Create route auth/reset_password
- Create route auth/forgot_password
- Generate and hash reset token
- Setup email template using mailgen
- Validate new Password
- Send reset and success emails

[Finishes #166272337]

EMacco committed:
feat-[166272337]: Forgot Password

- Write unit tests
- Document feature using swagger
- Create route auth/reset_password
- Create route auth/forgot_password
- Generate and hash reset token
- Setup email template using mailgen
- Validate new Password
- Send reset and success emails

[Finishes #166272337]

EMacco committed:
feat-[166272337]: Forgot Password

- Write unit tests
- Document feature using swagger
- Create route auth/reset_password
- Create route auth/forgot_password
- Generate and hash reset token
- Setup email template using mailgen
- Validate new Password
- Send reset and success emails

[Finishes #166272337]

EMacco committed:
feat-[166272337]: Forgot Password

- Write unit tests
- Document feature using swagger
- Create route auth/reset_password
- Create route auth/forgot_password
- Generate and hash reset token
- Setup email template using mailgen
- Validate new Password
- Send reset and success emails

[Finishes #166272337]

adex001 commented on server/controllers/users.js:
```diff
@@ -37,7 +48,13 @@ class UserController {
 
       const token = await Token.create(payload);
-      return res.status(201).json({ status: 'success', message: 'User created successfully', user: userExtractor(user, token) });
+      return res
```
you can use the Response class here. Response.success()

fidelisojeah replied:
@EMacco  Any reason why this feedback was ignored?

adex001 commented on server/controllers/users.js:
```diff
@@ -84,7 +101,13 @@ class UserController {
       };
       const token = await Token.create(payload);
-      return res.status(200).json({ status: 'success', message: 'User successfully logged in', user: userExtractor(user, token) });
+      return res
```
Response.success()

fidelisojeah replied:
@EMacco or this

adex001 commented on server/controllers/users.js:
```diff
@@ -138,20 +162,125 @@ class UserController {
-          message: 'Users and corresponding profiles retrieved successfully',
-          payload: users
-        });
+      return res.status(200).send({
```
Response.success() here too

adex001 reviewed (changes requested):
Nice job. Go ahead and implement the changes.

EMacco committed:
feat-[166272337]: Forgot Password

- Write unit tests
- Document feature using swagger
- Create route auth/reset_password
- Create route auth/forgot_password
- Generate and hash reset token
- Setup email template using mailgen
- Validate new Password
- Send reset and success emails

[Finishes #166272337]

EMacco committed:
feat-[166272337]: Forgot Password

- Write unit tests
- Document feature using swagger
- Create route auth/reset_password
- Create route auth/forgot_password
- Generate and hash reset token
- Setup email template using mailgen
- Validate new Password
- Send reset and success emails

[Finishes #166272337]

adex001 reviewed (dismissed):
Looks good to me. Great job!

emmygozi commented on server/tests/factory/user-factory.js:
```diff
@@ -26,6 +26,22 @@ const createTestUser = async ({
   return newUser;
 };
 
+const createNonActiveUser = async ({
```
If you can, how about you add a ```active: false``` to ```testUserNoArgumentPassed``` .
I think this function was created to simulate a non-active user. So we do not duplicate functions.

Otherwise, looks great to me. Good job.

GerrardE reviewed (dismissed):
Nice one. Good work with the documentation

EMacco committed:
feat-[166272337]: Forgot Password

 - Write unit tests
 - Document feature using swagger
 - Create route auth/reset_password
 - Create route auth/forgot_password
 - Generate and hash reset token
 - Setup email template using mailgen
 - Validate new Password
 - Send reset and success emails

[Finishes #166272337]

EMacco committed:
feat-[166272337]: Forgot Password

 - Write unit tests
 - Document feature using swagger
 - Create route auth/reset_password
 - Create route auth/forgot_password
 - Generate and hash reset token
 - Setup email template using mailgen
 - Validate new Password
 - Send reset and success emails

[Finishes #166272337]

fidelisojeah commented on server/helpers/mailer.js:
```diff
@@ -0,0 +1,82 @@
+
+const sendMail = ({ to, subject, message }) => {
+  const mailOptions = {
+    from: '"Author\'s Haven" <authorshaven@gmail.com>',
```
no hardcoded values.
Consider using environment variables for this

fidelisojeah commented on server/models/passwordreset.js:
```diff
@@ -1,3 +1,9 @@
+import bcrypt from 'bcrypt';
+
+const salt = process.env.SALT || 5;
+// eslint-disable-next-line radix
```
nope...don't do this 
```suggestion

```

fidelisojeah commented on server/models/passwordreset.js:
```diff
@@ -1,3 +1,9 @@
+
+const salt = process.env.SALT || 5;
+// eslint-disable-next-line radix
+const SALT_ROUNDS = parseInt(salt);
```
```suggestion
const SALT_ROUNDS = parseInt(salt, 10);
```
?

fidelisojeah commented on server/tests/factory/article-factory.js:
```diff
@@ -10,7 +10,6 @@ const createTestArticle = async (userId, { title, body }) => {
     slug: faker.lorem.slug(),
     title: title || faker.lorem.words(),
     body: body || faker.lorem.sentences(),
-    image: faker.image.imageUrl(),
```
any reason why this is removed though?

EMacco replied:
I think this happened when I was resolving a Merge conflict. I will put that back ASAP

adex001 reviewed (dismissed):
This is good. ~~You can make it better by removing your comments.

EMacco committed:
feat-[166272337]: Forgot Password

 - Write unit tests
 - Document feature using swagger
 - Create route auth/reset_password
 - Create route auth/forgot_password
 - Generate and hash reset token
 - Setup email template using mailgen
 - Validate new Password
 - Send reset and success emails

[Finishes #166272337]

adex001 reviewed (dismissed):
LGTM. Good job!

EMacco committed:
feat-[166272337]: Forgot Password

 - Write unit tests
 - Document feature using swagger
 - Create route auth/reset_password
 - Create route auth/forgot_password
 - Generate and hash reset token
 - Setup email template using mailgen
 - Validate new Password
 - Send reset and success emails

[Finishes #166272337]

EMacco committed:
feat-[166272337]: Forgot Password

 - Write unit tests
 - Document feature using swagger
 - Create route auth/reset_password
 - Create route auth/forgot_password
 - Generate and hash reset token
 - Setup email template using mailgen
 - Validate new Password
 - Send reset and success emails

[Finishes #166272337]

EMacco committed:
feat-[166272358]: Read History

- Write tests
- Create middleware to authenticate user
- Add article to read history of a user when the user fetches a single article
- Create route to fetch history for a user
- Document swagger

[Finishes #166272358]

EMacco committed:
feat-[166272358]: Read History

- Write tests
- Create middleware to authenticate user
- Add article to read history of a user when the user fetches a single article
- Create route to fetch history for a user
- Document swagger

[Finishes #166272358]

adex001 reviewed (dismissed):
Looks good to me.

fidelisojeah commented on server/controllers/users.js:
```diff
@@ -138,20 +159,137 @@ class UserController {
+
+      const token = await UserController.updateToken(user);
+
+      await sendForgotPasswordMail(token, email, user.get().username);
```
no need awaiting this.

Let it resolve in the background

EMacco replied:
I attempted to do that but it caused the tests for that feature to fail because the structure of the response was different from the structure of response from the Response.success()

Since I was not the one that worked on that ticket I communicated with @adex001 and he said it was fine that I can go ahead and resolve it. 

I changed the ones that did not affect the test tho.

fidelisojeah replied:
You are telling me
```suggestion
     sendForgotPasswordMail(token, email, user.get().username);
```
gives you issues?

EMacco replied:
No this works fine. I have already implemented it. 

I was referring to the feedback from @adex, but I have also implemented that too. Everything works.

EMacco committed:
feat-[166272337]: Forgot Password

     - Write unit tests
     - Document feature using swagger
     - Create route auth/reset_password
     - Create route auth/forgot_password
     - Generate and hash reset token
     - Setup email template using mailgen
     - Validate new Password
     - Send reset and success emails

    [Finishes #166272337]

EMacco committed:
feat-[166272337]: Forgot Password

 - Write unit tests
 - Document feature using swagger
 - Create route auth/reset_password
 - Create route auth/forgot_password
 - Generate and hash reset token
 - Setup email template using mailgen
 - Validate new Password
 - Send reset and success emails

[Finishes #166272337]

EMacco committed:
feat-[166272337]: Forgot Password

 - Write unit tests
 - Document feature using swagger
 - Create route auth/reset_password
 - Create route auth/forgot_password
 - Generate and hash reset token
 - Setup email template using mailgen
 - Validate new Password
 - Send reset and success emails

[Finishes #166272337]

EMacco committed:
feat-[166272337]: Forgot Password

 - Write unit tests
 - Document feature using swagger
 - Create route auth/reset_password
 - Create route auth/forgot_password
 - Generate and hash reset token
 - Setup email template using mailgen
 - Validate new Password
 - Send reset and success emails

[Finishes #166272337]

fidelisojeah commented on server/controllers/users.js:
```diff
@@ -138,20 +147,133 @@ class UserController {
+
+      // Get the user
+      const user = await User.findOne({ where: { email } });
+      if (!user) return Response.error(res, 404, 'User does not exist');
```
Your current implementation is a sort of security breach (You are giving a would be attacker information about if an email exists on our system or not)
consider using this instead
```suggestion
      if (!user) return Response.error(res, 400, 'Please use a valid reset link');
```

EMacco replied:
Makes sense. Thanks

fidelisojeah commented on server/controllers/users.js:
```diff
@@ -138,20 +147,133 @@ class UserController {
+      if (!tokenDetails) return Response.error(res, 401, 'Invalid Token');
+
+      const match = await bcrypt.compare(token, tokenDetails.get().resetPasswordToken);
+      if (!match) return Response.error(res, 401, 'Invalid Token');
```
```suggestion
      if (!match) return Response.error(res, 400, 'Please use a valid reset link');
```

fidelisojeah commented on server/helpers/mailer.js:
```diff
@@ -0,0 +1,85 @@
+
+config();
+
+const url = process.env.BASE_URL || 'https://kingsmen-ah-frontend-staging.herokuapp.com';
```
default this to localhost instead

fidelisojeah commented on server/helpers/mailer.js:
```diff
@@ -0,0 +1,85 @@
+      name,
+      intro: 'You are receiving this email because a password reset request for your account was received.',
+      action: {
+        instructions: 'Tap the button below to reset your customer account password. If you didn\'t request a new password, you can safely delete this email.',
```
Does eslint throw an error when you do this?
```suggestion
        instructions: `Tap the button below to reset your customer account password. If you didn't request a new password, you can safely delete this email.`,
```

EMacco replied:
Yes it does "Strings must use singlequote."

fidelisojeah replied:
Alright, makes sense.

EMacco committed:
feat-[166272337]: Forgot Password

 - Write unit tests
 - Document feature using swagger
 - Create route auth/reset_password
 - Create route auth/forgot_password
 - Generate and hash reset token
 - Setup email template using mailgen
 - Validate new Password
 - Send reset and success emails

[Finishes #166272337]

EMacco committed:
feat-[166272337]: Forgot Password

 - Write unit tests
 - Document feature using swagger
 - Create route auth/reset_password
 - Create route auth/forgot_password
 - Generate and hash reset token
 - Setup email template using mailgen
 - Validate new Password
 - Send reset and success emails

[Finishes #166272337]

EMacco committed:
feat-[166272337]: Forgot Password

 - Write unit tests
 - Document feature using swagger
 - Create route auth/reset_password
 - Create route auth/forgot_password
 - Generate and hash reset token
 - Setup email template using mailgen
 - Validate new Password
 - Send reset and success emails

[Finishes #166272337]

emmygozi reviewed (approved):
Looks good to me
